### PR TITLE
add `get_tagged_indices` method to `ADflowBuilder`

### PR DESCRIPTION
**Pull request by A-CGray**
## Purpose

Adds a `get_tagged_indices` method to `ADflowBuilder` as required for the multi-body ldtransfer implementation from https://github.com/OpenMDAO/mphys/pull/158

Also adds the zipper nodes to the surface coordinates passed from ADflow to MPhys so that we won't get dimension mismatches if you try using an overset mesh with zipper nodes in an aerostructural case. However, the derivatives of the forces at the zipper nodes do not seem to work so you should only do aerostructural cases with overset meshes if there are no overlaps on surfaces you want to include in the load and displacement transfer.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

I tested the functionality of `get_tagged_indices` using the CRM mesh from the test input files:

```python
# ==============================================================================
# Standard Python modules
# ==============================================================================
import os

# ==============================================================================
# External Python modules
# ==============================================================================
import numpy as np
from mpi4py import MPI
import openmdao.api as om
from mphys import Multipoint
from mphys.scenario_aerodynamic import ScenarioAerodynamic
from adflow.mphys import ADflowBuilder
from baseclasses import AeroProblem
import matplotlib.pyplot as plt
import niceplots

# ==============================================================================
# Extension modules
# ==============================================================================

MESH_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "input_files", "CRM_wbt_scalar_jst.cgns")

ADFLOW_OPTIONS = {
    # I/O Parameters
    "gridFile": MESH_FILE,
    "monitorvariables": ["cpu", "resrho", "resmom", "cl", "cd"],
    "surfaceVariables": ["cp", "vx", "vy", "vz", "mach"],
    "writeTecplotSurfaceSolution": True,
    "writevolumesolution": False,
    "writesurfacesolution": False,
    # Printing Parameters
    "setMonitor": True,
    "printTiming": False,
    # Load balancing parameters
    "loadImbalance": 0.1,
    "loadBalanceIter": 10,
    # Physics Parameters
    "equationType": "Euler",
}

COMM = MPI.COMM_WORLD

# --- Define a wing surface family ---
WING_SURFACES = ["inboard_up", "inboard_down", "outboard_up", "outboard_down"]
user_family_groups = {"wing": WING_SURFACES}

# --- Create the ADflow builder and get the underlying CFDSolver ---
builder = ADflowBuilder(ADFLOW_OPTIONS, user_family_groups=user_family_groups)
builder.initialize(COMM)
solver = builder.get_solver()

# --- Get the full surface coordinates vector and the wing surface coordinates ---
allCoords = solver.getSurfaceCoordinates(solver.meshFamilyGroup)
wingCoords = solver.getSurfaceCoordinates("wing")

# --- Now use get_tagged_indices to get the indices of the wing surface ---
wingIndicesFromTag = builder.get_tagged_indices(["wing"])
wingIndicesFromList = builder.get_tagged_indices(WING_SURFACES)

# --- Get what should be a second copy of the wing coordinates by indexing into the full surface coordinates ---
wingCoordsFromTag = allCoords[wingIndicesFromTag]
wingCoordsFromList = allCoords[wingIndicesFromList]

# --- Check that the two sets of coordinates are the same ---
for ii in range(COMM.size):
    if COMM.rank == ii:
        print(f"\nProc {ii}:")
        print(f"Number of surface coordinates on this proc, allWalls: {allCoords.shape[0]}, wing: {wingCoords.shape[0]}")
        tagWorks = np.array_equal(wingCoords, wingCoordsFromTag)
        listWorks = np.array_equal(wingCoords, wingCoordsFromList)
        print(f"builder.get_tagged_indices(['wing']) works? {tagWorks}")
        print(f"builder.get_tagged_indices(WING_SURFACES) works? {listWorks}")
    COMM.barrier()
```

The output from running `mpiexec -n 4 python MPhysTagsTest.py` is:

```
Proc 0:
Number of surface coordinates on this proc, allWalls: 975, wing: 720
builder.get_tagged_indices(['wing']) works? True
builder.get_tagged_indices(WING_SURFACES) works? True

Proc 1:
Number of surface coordinates on this proc, allWalls: 975, wing: 500
builder.get_tagged_indices(['wing']) works? True
builder.get_tagged_indices(WING_SURFACES) works? True

Proc 2:
Number of surface coordinates on this proc, allWalls: 649, wing: 220
builder.get_tagged_indices(['wing']) works? True
builder.get_tagged_indices(WING_SURFACES) works? True

Proc 3:
Number of surface coordinates on this proc, allWalls: 246, wing: 0
builder.get_tagged_indices(['wing']) works? True
builder.get_tagged_indices(WING_SURFACES) works? True
```

## Checklist

- [x] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Comment by A-CGray** [comment]
> How is the zipper mesh handled for these cases? I say we should figure out that part before doing anything multi-body like this.

Oh yeah I missed that argument, my assumption (based on [these lines](https://github.com/mdolab/adflow/blob/4ec8910a6f30425589c7c3be64f2fc635ef81ef0/adflow/mphys/mphys_adflow.py#L24-L26) in the mesh component) is that the zipper mesh is hidden from anything MPhys related, so I'll exclude it here too.

**Comment by A-CGray** [comment]
I've been discussing this with @anilyil and we're going to extend this PR to also make the MPhys wrapper's treatment of overset meshes more correct. Right now, the zipper mesh is ignored everywhere which would cause problems if we tried to do any load and displacement transfer with an overset mesh. I'll put this PR back in draft status until those features are ready to be reviewed.

**Comment by A-CGray** [comment]
I have been playing around this for a while now and can't get the coupled aerostructural adjoint to work with an overset mesh. I'm using the overset Onera M6 mesh from the MACH-Aero tutorial:
image

The primal solve works fine, but during the coupled-adjoint, the ADflow adjoint solves just stall out:

```
LN: LNBGS 0 ; 148.901745 1

==========================================================================
Calling ADflowForces.compute_jacvec_product for ONERAM6-AeroStruct problem
==========================================================================
-> Alpha... 3.060000 

================================================================
Calling ADflowSolver.solve_linear for ONERAM6-AeroStruct problem
================================================================
Solving linear in mphys_adflow
   0 KSP Residual norm 2.0981952276E-01
  10 KSP Residual norm 2.0888732710E-01
  20 KSP Residual norm 1.4499108329E-01
  30 KSP Residual norm 4.7088964066E-02
  40 KSP Residual norm 6.2618053613E-03
  50 KSP Residual norm 1.1332501409E-03
  60 KSP Residual norm 9.6055970949E-05
  70 KSP Residual norm 1.1093689153E-05
  80 KSP Residual norm 1.9332375995E-06
  90 KSP Residual norm 2.4096387003E-07
 100 KSP Residual norm 2.6584228888E-08

-> Alpha... 3.060000 
Updating states
-> Alpha... 3.060000 
Updating surface coords
Updating surface coords
-> Alpha... 3.060000 
Updating states
-> Alpha... 3.060000 
-> Alpha... 3.060000 
LN: LNBGS 1 ; 5.43831471 0.0365228407

-> Alpha... 3.060000 
Solving linear in mphys_adflow
   0 KSP Residual norm 2.7200195162E-01
  10 KSP Residual norm 2.7198510719E-01
  20 KSP Residual norm 2.7184132975E-01
  30 KSP Residual norm 2.6889573757E-01
  40 KSP Residual norm 2.0009278114E-01
  50 KSP Residual norm 1.1187536020E-01
  60 KSP Residual norm 1.0576940044E-01
  70 KSP Residual norm 1.0559526953E-01
  80 KSP Residual norm 1.0558818829E-01
  90 KSP Residual norm 1.0558794439E-01
 100 KSP Residual norm 1.0558793616E-01
.
.
.
Solving linear in mphys_adflow
   0 KSP Residual norm 2.3656381846E-02
  10 KSP Residual norm 2.3656381842E-02
  20 KSP Residual norm 2.3656381806E-02
  30 KSP Residual norm 2.3656381114E-02
  40 KSP Residual norm 2.3656366744E-02
  50 KSP Residual norm 2.3656353320E-02
  60 KSP Residual norm 2.3656352638E-02
  70 KSP Residual norm 2.3656352619E-02
  80 KSP Residual norm 2.3656352618E-02
  90 KSP Residual norm 2.3656352618E-02
 100 KSP Residual norm 2.3656352618E-02

-> Alpha... 3.060000 
Updating states
-> Alpha... 3.060000 
Updating surface coords
Updating surface coords
-> Alpha... 3.060000 
Updating states
-> Alpha... 3.060000 
-> Alpha... 3.060000 
LN: LNBGS 10 ; 0.0236738954 0.000158990047
```

And the resulting derivatives are waaaaay off:

```
-----------------
Total Derivatives
-----------------

+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| of '<variable>'           | wrt '<variable>' |   calc mag. |  check mag. |  a(cal-chk) |  r(cal-chk) | error desc |
+===========================+==================+=============+=============+=============+=============+============+
| 'cruise.aero_post.cd'     | 'aoa'            |  1.2854e-03 |  1.1910e-03 |  9.4432e-05 |  7.9289e-02 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.aero_post.cd'     | 'twist'          |  1.3683e-04 |  1.0550e-04 |  3.1329e-05 |  2.9696e-01 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.aero_post.cl'     | 'aoa'            |  4.0289e-02 |  3.6729e-02 |  3.5600e-03 |  9.6925e-02 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.aero_post.cl'     | 'twist'          |  1.2272e-03 |  4.2477e-05 |  1.1847e-03 |  2.7890e+01 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.aero_post.cmz'    | 'aoa'            |  2.3581e-03 |  1.9896e-03 |  3.6844e-04 |  1.8518e-01 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.aero_post.cmz'    | 'twist'          |  1.8820e-03 |  1.7589e-03 |  1.2317e-04 |  7.0027e-02 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.l_skin_ksFailure' | 'aoa'            |  9.8626e-01 |  9.6122e-01 |  2.5039e-02 |  2.6049e-02 |  >REL_TOL  |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
| 'cruise.l_skin_ksFailure' | 'twist'          |  4.6734e-01 |  4.5905e-01 |  8.2851e-03 |  1.8048e-02 |            |
+---------------------------+------------------+-------------+-------------+-------------+-------------+------------+
```

I've tried using more adjoint iterations and making the preconditioner ridiculously strong but neither really made any difference , additionally, I had no issues with the convergence of the adjoint in an aero-only case with the same mesh. Also, depending on the adjoint convergence tolerance I used, I'd sometimes have NaN's being returned from the ADflow MatVec routines which would crash everything.

I tried running a similar case with pyaerostructure and the derivatives there are fine, so I think this is something that can in theory be done.

@anilyil can we shift this PR back to just adding the `get_tagged_indices` method? I'm not planning to do any aerostructural stuff with overset meshes any time soon.

**Comment by A-CGray** [comment]
Hey @anilyil , I'm just getting back to this. We decided to give up on trying to make the wrapper work completely because it seems that ADflow has just never been able to correctly compute force derivatives for zipper meshes.

We did however agree to at least fix the array dimension mismatch that will currently occur if you use an overset mesh and try to compute forces. There are 2 ways we can do this:
- Include the zipper nodes in the surface coordinates (`x_aero0`) vector and the surface sensitivities
- Exclude the zipper nodes from the forces (`f_aero`) vector and the force sensitivities

Do you have any preference between these two? I think they're both similarly easy to implement.

**Comment by A-CGray** [comment]
I think this is finally actually ready for review now @anilyil @ArshSaja

File: adflow/mphys/mphys_adflow.py [hunk 1]
```diff
@@ -414,7 +439,7 @@ def setup(self):
         local_state_size = solver.getStateSize()
 
         self.add_input("adflow_vol_coords", distributed=True, shape_by_conn=True, tags=["mphys_coupling"])
-        self.add_output("adflow_states", distributed=True, shape=local_state_size, tags=["mphys_coupling"])
+        self.add_output("adflow_states", distributed=True, shape=local_state_size, tags=["mphys_coupling"], res_ref=1e7)
```
[the remark this answers is not in the document]
Comment (A-CGray) [reply adflow/mphys/mphys_adflow.py:442]: oh shoot, this wasn't supposed to be included in this PR, I was testing this out for my current aerostructural case and was planning to eventually open a PR to add it as a builder option.

**Comment by A-CGray** [comment]
> Can you please confirm that the compute method of ADflowForces includes the zipper nodes as well because the mapVector in the getForces call defaults to including zippers?

Yes, `getForces` ends with [this call](https://github.com/mdolab/adflow/blob/e082efa32d53e72bb4c63fb20676feeebdb9a41e/adflow/pyADflow.py#L3738) to `mapVector` for which `includeZipper` is True by default

**Comment by A-CGray** [comment]
@ArshSaja

**Comment by A-CGray** [comment]
@ArshSaja @lamkina This just needs one more approving review

**Comment by A-CGray** [comment]
@ArshSaja @lamkina This just needs one more approving review

**Comment by lamkina** [comment]
> @ArshSaja @lamkina This just needs one more approving review

Give me a day or two to test this on my case and then I'll approve if all goes well.

**Review by lamkina — APPROVED** [review]
I tested this on my case and it all worked as expected. LGTM.

**Pull request merged**